src/tls/test-server.c

```diff
@@ -345,6 +346,49 @@ test_no_tls_many_serial (TestCase *tc, gconstpointer data)
   g_assert_cmpuint (server_num_ws (), ==, 1);
 }
 
+static void
+test_tls_blocked_handshake (TestCase *tc, gconstpointer data)
+{
+  block_sigchld ();
+
+  pid_t pid = fork ();
+  if (pid == -1)
+    g_error ("fork failed: %m");
+
+  if (pid == 0)
+    {
+      /* child */
+      gint first_fd = do_connect (tc);
+      send_request (first_fd, "\x16"); /* start the TLS handshake */
+
+      /* Make sure the byte gets there before the next connection request */
+      sleep (1);
+
+      /* make sure we can do a second connection while the first one is
+       * blocked in the handshake
+       */
+      gint second_fd = do_connect (tc);
+      send_request (second_fd, "GET / HTTP/1.1\r\nHost: localhost\r\n\r\n");
+
+      /* wait 10 seconds for the reply */
+      struct pollfd pfd = { .fd = second_fd, .events = POLLIN };
+      g_assert_cmpint (poll (&pfd, 1, 10000), ==, 1);
+      close (second_fd);
```

Comment:
Couldn't this just use `assert_http()` here?

Reply:
No.  This is what I tried first, but this calls `do_request()` which iterates the server's mainloop from the calling thread (thus unblocking things).